runtime/src/accounts_hash.rs

```diff
@@ -869,37 +821,125 @@ impl AccountsHasher {
         min_index: usize,
         bin: usize,
         first_items: &'a mut Vec<Pubkey>,
```

Comment:
It seems that this `fn get_item` could be renamed to `next_item` for better clarity, as it is more like an `iterator.next()`, which not only returns the element point by `min_index`, but also updates the state of the iteration ...

Reply:
Comments are updated.

`get_index` is called from 2 locations, once per pubkey found (also for duplicates in multiple slots). It is highly coupled with its caller. It has been highly optimized for performance in this very slow algorithm.
`min_index` has to be a valid index in the relevant vectors, and the entry in `indexes` has to be valid in `sorted_data_by_pubkey`. I think the best assert here is the out of range panic we'll get if this were to somehow go wrong. Adding another assert would not give us much information. At various points I have put these values in a struct or other forms. We end up searching over `first_items` once per pubkey in the system (which can be 10B times). Even making it a ref instead of a copy was slower. On pop net simulations, we have 10k entries in the outer `sorted_data_by_pubkey`. This implies that `first_items` needs to be a distinct vector from `indexes` and `first_item_to_pubkey_division`.  Those later two could be combined into a single vector of struct or tuple, but we'd still have 2 vectors which would both have to match and they'd all have to still be consistent with the vec of mmapped cache files `sorted_data_by_pubkey`.
I've even considered making a subset version of `first_items` that only holds the last 8 bytes of a pubkey or something to use the cpu cache better.
Investigating this algorithm with 10k ancient append vecs, 200k bins, and 10B accounts could lead us to rethink or rewrite some of this algorithm further.
There is also an idea that we should break up the merkle trees into pubkey binned merkle trees so we could begin compressing into a merkle tree in parallel as soon as we know the FIRST pubkey in any bin (where bins are defined by the merkle tree binning and can be distinct from the bins the accounts hash algorithm uses now).
This accounts hash algorithm is a big ol pain in the rear at high accounts count.

Reply:
yeah. it is.

nit: can we bundle all the mutable parameters, i.e. first_items, indexes, frist_item_to_pubkey_division into a struct and pass the struct as a single mut ref?

Reply:
> bundle all the mutable parameters

I'd suggest we do this separately and profile it out.

Reply:
ok. sounds good to me.